await act(async () => ...)

threepointone opened the pull request:
TODO - 
- [x] ReactDOM, TestRenderer, NoopRenderer
- [x] tests for other renderers, I missed them last time around 
- [x] the warnings are placeholders, still thinking about what would be better, open to suggestions
- [x] implications on concurrent mode

![acting](http://www.reactiongifs.com/wp-content/uploads/2013/01/acting.gif)

(repasting from the group)

I hacked together an asynchronous version of `act(...)`, and it's kinda nice. 

You've seen the synchronous version -

```js
act(() => {
  // updates and stuff
})
// make assertions 
```

This still works, and gives the same warnings. But if you pass an `async` function - 
```js
await act(async() => {
  // updates and stuff
})
// expect commits and effects to be flushed 
// make assertions 
```
Neat! I set it up so if you *don't* `await` the result from `act`, a warning gets triggered (with `setImmediate`) to do so. That makes it a bit harder to have rogue async `act()` calls in the ether.

You can nest `act()` calls - 








```js
await act(async() => {
  // nest synchronous calls
  act(() => {
    // updates and such    
  })
  // as before, updates and effects are flushed 

  // make assertions 
  await sleep(500) // or for a promise to resolve, or whatever
  // more assertions maybe 

  // nest asynchronous calls too 
  await act(async () => {
    // mutations and such 
    // more awaits 
    // and maybe an event or two
  })
  // more assertions 
})
```

I implemented a cheap form of unrolling safety, so if a previous `act()` gets closed before any subsequent `act()` calls, a warning gets triggered. This should prevent most interleaving attempts, and maintain a tree-like shape of `act()` blocks. 

pros -
- works with async/await, solves oss problems cleanly 
- the sync api is preserved 
- the warning is preserved
- works with fake timers/fb 

cons - 
- can't guarantee batching after the first await in an act block (this will get better when concurrent?)
- less restrictive than the sync model, and starts to feel more opt-in than opt-out (eg- someone could just wrap their entire test with an `act()` call... which might be fine?)
- exposes a secret api on react dom to implement it, dunno how you feel about that

threepointone commented on packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js:
```diff
@@ -0,0 +1,294 @@
+    act = ReactTestUtils.act;
+  });
+
+  describe('sync', () => {
```
these are the ones moved in from TestUtils-test

threepointone commented on packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js:
```diff
@@ -0,0 +1,294 @@
+      );
+    });
+  });
+  describe('async', () => {
```
these are newly added

threepointone commented on packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js:
```diff
@@ -0,0 +1,294 @@
+        expect(div.innerHTML).toBe('0');
+        expect(ctr).toBe(1);
+      });
+
```
this might seem weird, but it matches user behaviour exactly - a flash of '0', followed by '1'.

threepointone commented on packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js:
```diff
@@ -0,0 +1,294 @@
+
+    it('it warns if you try to interleave multiple act calls', async () => {
+      spyOnDevAndProd(console, 'error');
+      (async () => {
```
we try to 'trick' act by awaiting its result, but not actually waiting for it.

threepointone commented on packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js:
```diff
@@ -0,0 +1,294 @@
+      });
+
+      await sleep(1000);
+      expect(console.error).toHaveBeenCalledTimes(1);
```
but nuh uh, we warn anyway (todo - match on actual warning message)

gaearon commented on packages/react-dom/src/test-utils/ReactTestUtilsAct.js:
```diff
@@ -0,0 +1,129 @@
+// a step counter when we descend/ascend from .act calls
+const actingScopeDepth = {current: 0};
+
+export default function act(callback: () => void | Promise<void>): Thenable {
```
What about the test renderer?

threepointone replied:
![image](https://user-images.githubusercontent.com/18808/52813917-c7fb3b80-3092-11e9-9365-d6b4b0806158.png)
working on it!

threepointone commented:
Assuming there’s no big problem with the approach, I’m going to close this pr later today and continue working on it in my branch. Please feel free to leave comments!

threepointone commented on packages/react-dom/src/__tests__/ReactTestUtils-test.js:
```diff
@@ -517,173 +515,4 @@ describe('ReactTestUtils', () => {
     ReactTestUtils.renderIntoDocument(<Component />);
     expect(mockArgs.length).toEqual(0);
   });
-
```
moved these tests into a separate file for .act tests

threepointone commented:
(nevermind, I'll leave it open)

threepointone commented:
spoke with dan, moving the `act(...)` logic into ReactFiberScheduler

threepointone commented:
The failing test is odd, investigating.

threepointone commented on packages/react-reconciler/src/__tests__/ReactHooks-test.internal.js:
```diff
@@ -886,11 +886,10 @@ describe('ReactHooks', () => {
 
     class Cls extends React.Component {
       render() {
-        act(() =>
```
we didn't need this act statement here

threepointone commented on packages/react-dom/src/test-utils/ReactTestUtils.js:
```diff
@@ -151,9 +147,35 @@ function validateClassInstance(inst, methodName) {
+      );
+    }
+    // then make it
+    actContainerElement = document.createElement('div');
```
lol this isn't needed anymore, but I'll keep the check.

gaearon commented on packages/react-reconciler/src/ReactFiberScheduler.js:
```diff
@@ -1827,9 +1827,59 @@ function scheduleWorkToRoot(fiber: Fiber, expirationTime): FiberRoot | null {
+  }
+
+  const result = batchedUpdates(callback);
+  if (result && result.then) {
```
We try to avoid `a && ...`, if you want to check `!== null && !== undefined` let's do that instead.

sebmarkbage replied:
https://github.com/facebook/react/blob/master/packages/react-reconciler/src/ReactFiberUnwindWork.js#L203-L205

Jessidhia commented on packages/react-reconciler/src/ReactFiberScheduler.js:
```diff
@@ -1827,9 +1827,59 @@ function scheduleWorkToRoot(fiber: Fiber, expirationTime): FiberRoot | null {
+  const result = batchedUpdates(callback);
+  if (result && result.then) {
+    // saving a few bytes without the typeof === 'function' check
+    return result.then(() => {
```
This will leak the semaphore if the `act` rejects. Should be in a `.finally` call (or one emulated with `.then(cb, cb)`.

threepointone replied:
Thanks, I’ll fix this and add a test

Jessidhia commented on packages/react-reconciler/src/ReactFiberScheduler.js:
```diff
@@ -1827,9 +1827,59 @@ function scheduleWorkToRoot(fiber: Fiber, expirationTime): FiberRoot | null {
+    // saving a few bytes without the typeof === 'function' check
+    return result.then(() => {
+      if (__DEV__) {
+        if (actingUpdatesScopeDepth - 1 > previousActingUpdatesScopeDepth) {
```
An option to support some interleaving is to instead only do the check if you were the call that first acquired the semaphore, but I'm not 100% sure that is safe.

threepointone replied:
I don’t think we want to support interleaving, do you have an example of when you’d want to interleaved test interactions?

threepointone commented:
@jessidhia 
> I'm not sure how "(this will get better when concurrent?)"
This is an incomplete implementation, I was only trying to get the shape of the api right and pass tests. Still discussing/sketching out it would look like.

thysultan commented:
Was there any reason await-ables / callbacks where not considered, i.e:

```js
function App(props) {
 React.useEffect(props.callback);
 return null;
}

const container = document.createElement('div');

let called = false;

ReactDOM.render(<App callback={() => { called = true; }}/>, container, (current) => {
  expect(called).toBe(true);
});

// or

await ReactDOM.render(<App callback={() => { called = true; }}/>, container);
expect(called).toBe(true);
```

That is instead of trying to force async logic to become sync, why not improve async and invoke the render callback when all of this work is done(whenever this is), and return a thenable from `render` for good measure to allow awaiting it.

gaearon commented:
This API is not for `ReactDOM.render` specifically.

sebmarkbage commented on packages/react-test-renderer/src/ReactTestRenderer.js:
```diff
@@ -564,40 +560,48 @@ const ReactTestRendererFiber = {
-        );
+    // createReactNoop.js and ReactTestUtils.js
+    const result = actedUpdates(callback);
+    if (result && result.then) {
```
We already have a definition for how to test a thenable:
https://github.com/facebook/react/blob/master/packages/react-reconciler/src/ReactFiberUnwindWork.js#L203-L205

sebmarkbage commented on packages/react-test-renderer/src/__tests__/ReactTestRenderer-test.js:
```diff
@@ -29,6 +37,54 @@ describe('ReactTestRenderer', () => {
   });
 
+  describe('act', () => {
+    it('should work', () => {
```
better name

threepointone commented:
I cleaned up what I could, and I'm down to 2 real 'issues' 
- I spotted some odd behaviour with ReactTestRenderer, where effects don't seem to get flushed unless a promise gets resolved. similar tests pass with ReactDOM/TestUtils, and the NoopRenderer. I'm investigating, but it's reproducible with the failing test
- like @sebmarkbage said,  I'll remove this from the scheduler and move to testutils. It *should* be doable.

threepointone commented on packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js:
```diff
@@ -0,0 +1,309 @@
+  });
+});
+
+// todo - errors are caught as expected
```
```suggestion
```

threepointone commented on packages/react-test-renderer/src/__tests__/ReactTestRenderer-test.js:
```diff
@@ -29,6 +29,99 @@ describe('ReactTestRenderer', () => {
+        }
+        let root;
+        await ReactTestRenderer.act(async () => {
+          // this test will fail
```
this bit is the odd behaviour

acdlite commented on packages/react-noop-renderer/src/createReactNoop.js:
```diff
@@ -873,41 +868,52 @@ function createReactNoop(reconciler: Function, useMutation: boolean) {
+                // todo - a better warning here. open to suggestions.
+              );
+            }
+          }, 0);
```
Using a timeout isn't ideal because it only works if the timers run. Let's figure out something better. At the very least, we should use a microtask instead of a timer event.

(In retrospect, maybe what we should have done is wait to call the callback until the thenable is awaited. Then `act` wouldn't do anything _unless_ you await the promise. But I don't think there's a way to change that without breaking everyone's existing tests.)

acdlite replied:
Yeah I think if you make this a microtask (Promise.then()) instead of `setTimeout` then it's good enough. Jest will still fail even if nothing is awaited.

acdlite replied:
Welp, never mind, that only works because of the way we set up Jest in the React repo.

acdlite replied:
Still, microtask is probably our best option. We can looking into having Jest always wait a microtask at the end each test.

threepointone replied:
the problem with the microtask is that it resolves before any further created promises resolve, triggering the warning every time.

acdlite replied:
How about two microtasks? :D

https://github.com/facebook/react/blob/a477f6d2fa056fdb54bf8a58037f7c225b4fb997/packages/react-noop-renderer/src/createReactNoop.js#L841-L848

acdlite commented on packages/react-test-renderer/src/ReactTestRenderer.js:
```diff
@@ -564,40 +560,52 @@ const ReactTestRendererFiber = {
+          if (__DEV__) {
+            warningWithoutStack(
+              false,
+              // todo - well... why not? maybe this would be fine.
```
Yeah this is the opposite of what we want. Should always await.

acdlite commented on packages/react-test-renderer/src/ReactTestRenderer.js:
```diff
@@ -564,40 +560,52 @@ const ReactTestRendererFiber = {
+      let called = false;
+      if (__DEV__) {
+        setTimeout(() => {
+          if (!called) {
```
We should only fire this warning if we're in concurrent mode, I think. Otherwise it's a breaking change. So we'll need some way to detect if the callback scheduled concurrent work.

acdlite replied:
Ah, or I guess not because we currently warn if the callback returns a promise. Never mind.

acdlite replied:
Although it does seem like, ideally, we would enforce that you _always_ await `act`. So that microtasks have a chance to fire. Unfortunately that's a breaking change :/

threepointone replied:
just to clarify how we resolved this - I'll verify that jest won't swallow warnings when this is fired 'out of band' in sync tests.

threepointone commented:
Alright, so I've updated this PR - 

- please have a look at ReactTestRendererAct-test - "should work with async/await" I got the test to pass, but I'm not certain that's the expected behaviour (it kinda looks like it is though?)
- it now uses the ReactFiberScheduler::flushPassiveEffects (instead of the respective hacks)
- it now recursively flushes effects, fixing #14954 
- but that also means I *cannot* move this code outside ReactFiberScheduler :( cc @sebmarkbage 

I'm looking into how I can save some bytes in prod, but this seems to be functionally ready. I tested with a variety of OSS use cases and it should be good. I'm going to update my examples repo with this and add a bunch more usecases there, and also rewrite the (admittedly wrong) details in it.

acdlite commented on packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js:
```diff
@@ -0,0 +1,326 @@
+let ReactTestUtils;
+let act;
+
+jest.useRealTimers();
```
I don't get it. Why can't you use jest's fake timers?

threepointone replied:
OSS can’t/don’t use fakeTimers reliably, so I want to run tests in an environment as close to theirs as possible. (Specifically for the async/await test, which started the need for this pr)

acdlite replied:
Can you move this to the specific test(s) where you rely on native timers then?

threepointone replied:
Missed erplying to this. It's all the tests, except the ones that use fake timers. I want to be explicit about using fake timers in this suite.

sebmarkbage replied:
It is very hard to see if you cover all case for both environments. By default I expect there to be one file for each environment and every test mirrored in each file so that I can see that you've covered all cases in each environment and how you expect any case to be ported.

If some test needs to test something specific that is only relevant in one environment sure, but I expect code to mostly work the same in either environment.

Therefore splitting these into two separate files with a clear name in the file name which environment you're testing would be good.

threepointone replied:
That's fair, I could do that.

acdlite commented on packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js:
```diff
@@ -0,0 +1,326 @@
+        ReactDOM.render(
+          <App
+            callback={() => {
+              called = true;
```
This seems nitpicky but by convention in our tests we push things into a log and then assert on the log. That way if, say, the callback is called twice instead of once, the test will fail.

threepointone replied:
Good point, I’ll make it so

threepointone replied:
I fixed this, but I found a bunch of other usages of this pattern in the tests haha https://github.com/facebook/react/search?q=%22called+%3D+true%22&type=Code

acdlite replied:
Now we have fewer! :D

acdlite commented on packages/react-test-renderer/src/__tests__/ReactTestRendererAct-test.js:
```diff
@@ -0,0 +1,105 @@
+
+      expect(root.toJSON()).toEqual('1');
+
+      // await ReactTestRenderer.act(async () => {
```
What's going on here?

threepointone replied:
there's a difference in behaviour between the TestRenderer and DOM, and I thought it shouldn't. A similar test in ReactDOM works as I'd expected, but I guess it just works that way though. I can make a separate PR showing the difference. deleted this wall of text here.

threepointone commented:
I've addressed the feedback, working separately on my examples repo. Can I get a review for this plsthx :)

acdlite commented on packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js:
```diff
@@ -0,0 +1,329 @@
+          ReactDOM.render(<App />, el);
+        });
+
+        await sleep(500);
```
So the reason I don't like using native timers in all of these tests is that now it takes at least 500ms before this test can finish. Whereas if you use fake timers, it's CPU bound.

threepointone replied:
I can reduce the timer length, but this is specifically to test when we can't use fake timers reliably.

acdlite replied:
It's fine to use real timers in one or two of these tests but doing it in all of them seems like overkill especially since most people who use timers are going to use the fake timers, anyway.

acdlite commented on packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js:
```diff
@@ -0,0 +1,329 @@
+        }
+        const container = document.createElement('div');
+
+        act(() => {
```
These `act` calls are not being awaited :(

acdlite commented on packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js:
```diff
@@ -0,0 +1,329 @@
+          ReactDOM.render(<App />, container);
+        });
+        act(() => {
+          jest.runAllTimers();
```
If you replace the `setTimeout` in the effect with `sleep(200).then(() => setToggle(1))` then the test will fail, because the update happens in a microtask. That's why it's important that you always await `act`. We need another test case that shows that.

threepointone replied:
I'm not sure that's true with fakeTimers, but I'll verify.

acdlite replied:
Even if it's not true for fake timers (though I think it is) it would be true for other types of events, like network.

gaearon replied:
>most people who use timers are going to use the fake timers, anyway.

I don't believe this to be the case in open source. I've mostly seen people being wary of mocking timers for different reasons (often emotional ones like "doesn't feel right") and we need to meet them where they are.

threepointone replied:
so I tried, and - 
- in our setup (and likely oss too), this will trigger the warning. the fix for this is simple - use the async version to catch microtasks like so 
```jsx
await act(async () => {
  jest.runAllTimers()
})
```
- I'd assumed we had a setup similar to fb's, where the changed code *will not* trigger the warning (I just tested)

I'm updating the test to show this.

threepointone replied:
this is now better since we wait for a whole task.

acdlite commented on packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js:
```diff
@@ -0,0 +1,329 @@
+    it('warns if you try to await an .act call', () => {
+      expect(act(() => {}).then).toWarnDev(
+        [
+          'Do not await the result of calling act(...) with sync logic, it is not a Promise.',
```
I don't think we want this anymore, right? Should always await.

threepointone replied:
Is that what we discussed? are we making a breaking change here? (actually asking, I thought we decided not to)

acdlite replied:
I don't think we treat new warnings as a breaking change. We introduce warnings in minors all the time. It's a bit churny, but I think it's worth it. I might not have thought through all the implications though so perhaps we should discuss it during our team meeting.

gaearon replied:
This seems like it would be rather churny given it was prominent in the blog post and others already integrated it into wrapper libraries (like `react-testing-library`). The bigger problem is people would have to convert all their React tests to be async, right? That's not always ideal.

Btw speaking of which, should we reach out to testing libraries to ask them to proxy the returned value?

kentcdodds replied:
Yes, please let's not force all tests to be async 😬

threepointone replied:
I don't think we need to await all acts. In sync mode, it works fine with sync-y code. if tests queue promises, they can use the async version (and await on that). now, tests written for sync mode _could_ fail in concurrent mode, at which point we have one of two options - rejig the test setup to match fb's so we can resolve promises immediately, or use the async version of act. both are reasonable choices depending on your constraints.

acdlite replied:
It doesn't have anything to do with concurrent mode. The motivation for requiring `await` is it flushes pending promise resolutions.

For example, if you call `jest.advanceTimers(ms)`, that will fire timer events, but the promises won't resolve, which leads to subtle differences in how your component will work in a real environment. (See: https://github.com/facebook/react/pull/14853#discussion_r263034445). Like if a promise handler includes additional setStates, the test will exit before those have a chance to render, potentially leading to false negatives.

Maybe the downsides of using async/await in all tests outweigh the benefits but we shouldn't dismiss it without considering this pitfall.

acdlite replied:
Another way to think about this is that microtasks are another kind of "passive event." Ideally, `act` should always flush microtasks at the end for the same reason it flushes passive events at the end.

kentcdodds replied:
I'd just like to _only_ have to `await` in my test if my component actually is doing something async (typically with promises). Otherwise I'd like to keep my test sync is all :)

acdlite replied:
I understand the reasons why it's annoying but nobody seems to be addressing my point about dangling effects in promise handlers.

gaearon replied:
I think it's a reasonable compromise to use async `act` in docs and blog posts but not warn for existing usage. People who discover issues or care will learn to convert to async, and eventually the knowledge will propagate.

threepointone replied:
a side effect of making awaiting act mandatory, is people wouldn't be able to make synchronous assertions about their UI (when nesting inside an async act, for example). I think we want to keep this behaviour. marking this bit as resolved, unless y'all think differently.

acdlite commented on packages/react-dom/src/test-utils/ReactTestUtils.js:
```diff
@@ -389,58 +387,68 @@ const ReactTestUtils = {
+      let called = false;
+      if (__DEV__) {
+        // eslint-disable-next-line no-undef
+        Promise.resolve()
```
Technically we don't require promise support. Since this is just a warning, wrap it in an existence check.

threepointone replied:
Getting a particularly odd error when I put in the existence check (see the CI error.) posted about it internally, let's see.

acdlite replied:
Which CI error? And I can't find the internal post. Send to me via Work Chat? :D

threepointone replied:
alright, the current version seems to appease both lint and flow. marking this resolved.

acdlite commented on packages/react-noop-renderer/src/createReactNoop.js:
```diff
@@ -706,41 +701,52 @@ function createReactNoop(reconciler: Function, useMutation: boolean) {
+        let called = false;
+        if (__DEV__) {
+          // eslint-disable-next-line no-undef
+          Promise.resolve()
```
Same here

acdlite commented on packages/react-reconciler/src/ReactFiberHooks.js:
```diff
@@ -1026,16 +1026,14 @@ function updateMemo<T>(
   // $FlowExpectedError - because requirements don't give a damn about your type sigs.
   if ('undefined' !== typeof jest) {
-    shouldWarnForUnbatchedSetState = true;
+    shouldWarnForUnactedSetState = true;
```
I know this isn't new but this is the first time I've noticed it. How does this work for non-Jest test runners?

threepointone replied:
Everything 'works', except we don't warn when you don't use act around an update. We considered other options (NODE_ENV=test etc), but this seemed most reasonable for now.

gaearon commented on packages/react-test-renderer/src/ReactTestRenderer.js:
```diff
@@ -553,60 +549,57 @@ const ReactTestRendererFiber = {
+    ) {
+      let called = false;
+      if (__DEV__) {
+        if (typeof global.Promise !== 'undefined') {
```
This will crash outside of Node right? There's no `global` elsewhere. Not sure if people run test renderer in the browser but they might.

gaearon replied:
If you're just trying to appease flow, add a suppression. These types aren't used anywhere anyway.

threepointone commented:
I'm spending some time today trying to flush promises at the end of act.

threepointone commented on packages/react-reconciler/src/ReactFiberScheduler.js:
```diff
@@ -1824,9 +1824,93 @@ function scheduleWorkToRoot(fiber: Fiber, expirationTime): FiberRoot | null {
+  function flushEffectsAndMicroTasks() {
+    return new Promise((resolve, reject) => {
+      flushPassiveEffects();
+      setImmediate(() => {
```
working on an xplat version of this asap (probably postMessage)

threepointone replied:
(done)

threepointone commented on packages/react-reconciler/src/ReactFiberScheduler.js:
```diff
@@ -1824,9 +1824,93 @@ function scheduleWorkToRoot(fiber: Fiber, expirationTime): FiberRoot | null {
+      }
+    }
+  }
+  function flushEffectsAndMicroTasks() {
```
I considered doing this with callbacks, but this reads a bit better.

threepointone commented:
I chatted with @acdlite with some clarifications and think this is ready for a review. I have an idea to reduce bytes, I'm working on that right now.

acdlite commented on packages/react-dom/src/test-utils/ReactTestUtils.js:
```diff
@@ -389,58 +387,68 @@ const ReactTestUtils = {
+        // eslint-disable-next-line no-undef
+        Promise.resolve()
+          .then(() => {})
+          .then(() => {
```
Shouldn't need this extra jump anymore

threepointone replied:
still do, since the thenable that triggers the warning may be called after one microtask

acdlite commented on packages/react-reconciler/src/ReactFiberScheduler.js:
```diff
@@ -1824,9 +1824,106 @@ function scheduleWorkToRoot(fiber: Fiber, expirationTime): FiberRoot | null {
+
+  function flushEffectsAndMicroTasks() {
+    // eslint-disable-next-line no-undef
+    return new Promise((resolve, reject) => {
```
Use a thenable instead of a promise

threepointone replied:
I tried, but it still needs the microtask before running flushPassiveEffects again, making this hairier. I'm moving this out of ReactFiberScheduler (into TestUtils/TestRenderer/createReactNoop), so it's... fine, maybe?

threepointone replied:
consider also that this codepath is only called with the async version of act, which means they probably have Promise defined on the page

acdlite commented on packages/react-reconciler/src/__tests__/ReactNoopRendererAct-test.js:
```diff
@@ -0,0 +1,63 @@
+const ReactNoop = require('react-noop-renderer');
+const Scheduler = require('scheduler');
+
+describe('ReactNoop.act()', () => {
```
Because the macrotask thing is so subtle, we need tests that correspond each of the platforms where you might run tests.

- Node + jsdom (or other fake dom)
- Node without jsdom
- React Native
- No setImmediate
- Polyfilled setImmediate

acdlite replied:
Polyfilling setImmediate is very common and the semantics are non-standard. It's supposed to be a macrotask but I don't know if we can rely on that.

acdlite replied:
For example, this part indicates to me that `act` will fail if neither `setImmediate` nor `MessageChannel` is defined:

https://github.com/facebook/react/pull/14853/files#r265820721

(Maybe I'm missing something, but regardless, it would be nice to have a test that verifies this.)

threepointone replied:
> we need tests that correspond to each of the platforms where you might run tests.

do you still feel strongly about this? now that - 
- we use the non-mocked setImmediate from node
- we aren't trying to use anything beyond MessageChannel (and warn if you try in a browser that doesn't have it)

acdlite replied:
They don't need to be super elaborate tests, just a sanity check that this won't insta-fail in those environments so that the next time someone touches this code they don't forget to consider those cases.

threepointone replied:
btw, the decision was to let it insta fail when setImmediate/MessageChannel aren't available (so, 'old' IE). I have a backup plan in case that's needed, but letting it be. I'll set up tests for the rest.

kentcdodds replied:
I'm personally fine with that plan so long as there's a useful error message that explains what's going on clearly please. Because I just know I'll get issues on react-testing-library otherwise.

threepointone replied:
https://github.com/facebook/react/pull/14853/files#diff-08838400dbefc5107cea81326e3d9847R176

kentcdodds replied:
Perfect 💯

acdlite commented on packages/react-test-renderer/src/__tests__/ReactTestRendererAct-test.js:
```diff
@@ -0,0 +1,83 @@
+  });
+
+  describe('async', () => {
+    it('should work with async/await', async () => {
```
This test doesn't make much sense to me. Can you write this test in the way you expect people to write their product tests?

threepointone replied:
deal

acdlite commented on packages/react-test-renderer/src/__tests__/ReactTestRendererAct-test.js:
```diff
@@ -0,0 +1,83 @@
+      let root;
+      ReactTestRenderer.act(() => {
+        root = ReactTestRenderer.create(<App />);
+        Promise.resolve();
```
What is this line for?

threepointone replied:
I'll rewrite this entire test, it's now a lot better with the way we 'wait' for stuff, and the user won't have to do any manual `Promise.resolve()` statements.

acdlite commented on packages/react-test-renderer/src/__tests__/ReactTestRendererAct-test.js:
```diff
@@ -0,0 +1,83 @@
+        Promise.resolve();
+      });
+      // The TestRenderer needs a micro task before it advances
+      // but we can catch it with act()
```
I don't get this comment

threepointone replied:
rewrote this test completely

acdlite commented on packages/react-reconciler/src/ReactFiberScheduler.js:
```diff
@@ -1824,9 +1824,106 @@ function scheduleWorkToRoot(fiber: Fiber, expirationTime): FiberRoot | null {
+
+// we 'wait' for queued microtasks to resolve (after flushing effects)
+// using either setImmediate when available, or postMessage
+const nextTick =
```
Nit: `nextTick` might not be the best tick because in Node it behaves like a microtask, not a macrotask. But you specifically want this to resolve after a macrotask.

threepointone replied:
how does enqueueTask sound?

acdlite commented on packages/react-reconciler/src/ReactFiberScheduler.js:
```diff
@@ -1824,9 +1824,106 @@ function scheduleWorkToRoot(fiber: Fiber, expirationTime): FiberRoot | null {
+  typeof setImmediate !== 'undefined'
+    ? setImmediate
+    : callback => {
+        const channel = new MessageChannel();
```
What if `MessageChannel` is not defined?

threepointone replied:
so this is in older browsers (ie9 and below and some others). the annoying scenario is when someone's using fakeTimers + async act, and running their tests across browser. we can't use setTimeout in this case because it would require manually advancing after each act call. So the challenge is to delay by one macrotask, but without using any of the usual suspects - setTimeout, setImmediate, etc. 

Here's a super dirty version of a workaround
```jsx
function nextTick(callback){
  let promise = Promise.resolve()
  let ctr = 0
  while(ctr < 100){
    ctr++
    promise = promise.then(() => {})
  }
  return promise.then(() => {
    callback()
  })
};
```
This isn't satisfying; it's slow, and there are obvious edge cases. 

Another option is to create a dummy element (like an img/iframe/link) and try to load an inlined resource. something like this -
```jsx
function nextTick(callback){
  const img = new Image()
  img.onload = callback
  img.src = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkYAAAAAYAAjCB0C8AAAAASUVORK5CYII='
}
```
this one's less ickier. I tested it, and it "works". Thoughts?

gaearon replied:
I don't think we care that TestUtils work in IE9.

threepointone replied:
so... should I not have the fallback, or just throw an error? our 'support' usually involves just adding polyfills, and this is a scenario where that might be enough (it'll be odd to ask folks to install a MessageChannel polyfill to pass their tests, but sure.)

gaearon replied:
If we can polyfill it easily (and the polyfill lives outside production code path) then it seems ok. Otherwise — I'm not sure it's reasonable that everyone using ReactDOM in prod pays the cost for the hypothetical users who run TestUtils in IE9.

threepointone replied:
I'm moving all this logic into TestUtils, so it won't add to ReactDOM

sebmarkbage commented on packages/react-reconciler/src/ReactFiberScheduler.js:
```diff
@@ -1824,9 +1824,106 @@ function scheduleWorkToRoot(fiber: Fiber, expirationTime): FiberRoot | null {
+// using either setImmediate when available, or postMessage
+const nextTick =
+  typeof setImmediate !== 'undefined'
+    ? setImmediate
```
What if `setImmediate` is mocked? [jest mocks it](https://github.com/facebook/jest/blob/master/packages/jest-fake-timers/src/jestFakeTimers.ts)

Code like this would pick up the mocked version:

```js
jest.useFakeTimers();
let ReactDOM = require('react-dom');
```

In that setup we wouldn't flush any work at the end of act until timers are ran. This is bad. We should make this just work.

Is this the reason you have to do weird things like call useRealTimers and call runAllTimers at random places?

sebmarkbage replied:
If we can somehow instead `require('timers').setImmediate` from the Node environment, we can get to the unmocked version.

```
let setImmediate;
try {
  let r = require; // trick packagers not to bundle this stuff.
  setImmediate = r('timers').setImmediate;
} catch (x) {
  // try one of the other choices
}
```

threepointone replied:
> Is this the reason you have to do weird things like call useRealTimers and call runAllTimers at random places?

(notes for posterity's sake)

I need to `useRealTimers()` because we use fake timers by default, but OSS folks don't, so I need to explicitly disable them for our tests. 
There are 4 'scenarios' -
- fake timers, with transpiled environments (alá fb): in this setup, the async version of act is barely (never?) used, and peeps use `runAllTimers()` to 'advance' time. of note, the above codepath is never used. this is typical code when so - 
```jsx
act(() => {
  jest.runAllTimers();
  // advance time and capture 
  // updates and effects
});
``` 
- real timers, common js presets: closer to what OSS has. in this scenario, both the sync and async versions work as expected, no real surprises. you would see code like this -
```jsx
await act(async () => {
  doSomething()
  await waitForElement(() => getByTestId('greeting-text'))
}) 
```
- real timers, transpiled environments: similar to the previous one, both sync/async versions will work as expected. 
- fake timers, common js presets: this is the default setup we have in the react repo, and where tests with cascading promises would fail (it'll pass the first tick with `await act(async () =>   jest.runAllTimers();`, but thats not sufficient). your proposed hack fixes this! for browsers without MessageChannel, how do you feel about this hack? https://github.com/facebook/react/pull/14853/#discussion_r265993434

sebmarkbage replied:
> act is barely (never?) used, and peeps use runAllTimers() to 'advance' time. of note, the above codepath is never used

I'm confused about the expectation here. I think this is what we explicitly __don't__ want to have happen.

Best practice is to not call runAllTimers inside act because act only means that the event turns have happened, not that time has elapsed. An event turn can execute infinitely quickly.

For example I often want do:

```js
await act(() => {
  setState();
});
// expect on the first commit
await act(() => {
  runAllTimers();
});
// expect on subsequent commits that happened after timers set up in effects fired
```

Similarly sometimes you want to call timers without actually flushing any rendering work.

```js
await act(() => {
  setState();
  runAllTimers();
  // the set state above is still not flushed because it might have taken awhile to render
});
// expect on the commit
```

Is this also your expectation?

threepointone replied:
I think we're on the same page? The examples you just gave should still work. I simply meant that *in an fb-like-environment*, there isn't much use (yet!) for the *async version of act*, since - 
```jsx
act(() => {
  runAllTimers();
});
```
will effectively do the same thing as -
```jsx 
await act(async () => {
  runAllTimers();
});
```
(which is to resolve queued promises and timers.)

threepointone commented:
addressed all the feedback I could. now working on https://github.com/facebook/react/pull/14853#discussion_r265819394 to run some sanity tests under combinations of environments to make sure it all works as expected.

acdlite commented on packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js:
```diff
@@ -0,0 +1,403 @@
+          ReactDOM.render(<App />, container);
+        });
+        await act(async () => {
+          jest.runAllTimers();
```
As a follow-up item, we should talk to Jest about a version of fake timers that encourages/requires you to wait for a new task. I'm guessing they haven't thought much about this problem because Facebook polyfills promises and transpiles async/await.

threepointone replied:
I spoke to a couple of folks around here, and they mentioned they're looking at using https://nodejs.org/api/async_hooks.html

acdlite commented on packages/react-dom/src/test-utils/ReactTestUtils.js:
```diff
@@ -45,12 +41,20 @@ const [
 
 let hasWarnedAboutDeprecatedMockComponent = false;
 
+let globalPromise = null;
```
Explain to me again why we need to use the built-in promises? I still don't get why a thenable isn't sufficient.

sebmarkbage replied:
Maybe point to a test that fails if you use a thenable?

threepointone replied:
I must have been doing something wrong. I managed to get this to work.

acdlite commented on packages/react-dom/src/test-utils/ReactTestUtils.js:
```diff
@@ -390,58 +444,70 @@ const ReactTestUtils = {
+    ) {
+      let called = false;
+      if (__DEV__) {
+        // $FlowFixMe - if we've reached this far, we're certain Promise is available
```
This is in a dev block, anyway, and Promise will almost certainly always be defined, so it can use an existence check. The bar for using $FlowFixMe needs to be really high and I don't think it's necessary in this case.

Alternatively we should decide that Promise is a required global and update our ESLint config accordingly.

threepointone replied:
the problem with an existence check, is that flow infers that Promise might not exist as a global, and it throws weird errors. sent you a link privately to internal post on the same.

acdlite replied:
That error doesn't seem to relevant to me. It says Promise is not a polymorphic type, which is more a problem with Flow's built-in type defs. But that shouldn't matter here because you don't need to type the resolved value; you only need to call `Promise.resolve().then`:

https://flow.org/try/#0AoJw9gtglgzgpgOhHGYA2A3OAKAlAgFwAs4A7PIA

threepointone replied:
It poisons any function with a Promise in its type annotation. I'll keep trying to fix this, but fair warning I already tried a bit and kept hitting walls. 
I would heartily +1 updating the eslint config. It seems reasonable at this point.

acdlite replied:
You don't need to use `Promise` in the types of any of these functions though. You use `Thenable<T>`.

threepointone replied:
Alright, I must have been doing something wrong earlier. I "fixed" this.

acdlite commented on packages/react-dom/src/test-utils/ReactTestUtils.js:
```diff
@@ -152,9 +156,59 @@ function validateClassInstance(inst, methodName) {
+    if (didWarnAboutMessageChannel === false) {
+      didWarnAboutMessageChannel = true;
+      warningWithoutStack(
+        typeof MessageChannel !== 'undefined',
```
Wrap in a DEV check plz (I know we have a Babel plugin that does this automatically but we try to be explicit and I'd like to get rid of it in the future.)

threepointone replied:
ack fair, missed it.

acdlite commented on packages/react-reconciler/src/ReactFiberScheduler.js:
```diff
@@ -1824,9 +1824,90 @@ function scheduleWorkToRoot(fiber: Fiber, expirationTime): FiberRoot | null {
+  callback: () => void | Promise<void>,
+  runCallbackUntilPredicateFails: (
+    callback: () => void,
+    predicate: () => boolean,
```
`runCallbackUntilPredicateFails` isn't a public API, right? Maybe I missed a conversation about this?

acdlite replied:
I had expected you'd solve this problem with something like `doesHavePendingPassiveEffects`. Why is this preferable? Remember this is a temporary workaround and the real solution is to have a custom testing build of React DOM.

threepointone replied:
Have a look now? Do you think I should still export `doesHavePassiveEffects` instead of passing it as a callback?

acdlite commented:
Why does making it a callback improve minification? In terms of size, imports and exports between internal modules are zero cost because of Rollup.

threepointone commented:
the callback itself doesn’t reduce size, it only kept it down to one export. The bit that mattered was not having to include .then chaining, especially since act doesn’t do anything else in prod besides just calling the callback. happy to export doesHavePassiveEffects as well btw, but didn’t want to stop my refactor in the meanwhile.

acdlite commented on packages/react-dom/src/test-utils/ReactTestUtils.js:
```diff
@@ -390,58 +438,120 @@ const ReactTestUtils = {
-          addendum = ' You returned: ' + result;
+    let thenable;
+    actedUpdates((doesHavePassiveEffects, onDone) => {
+      const result = ReactDOM.unstable_batchedUpdates(callback);
```
If it'a already wrapped in batched updates then it shouldn't need to be wrapped in batchedUpdates, too

threepointone replied:
Hmm? I moved the batchedUpdates call from FiberScheduler to here

acdlite replied:
If it's already wrapped in `actedUpdates`, I meant.

> I moved the batchedUpdates call from FiberScheduler to here

Why though? Remember that eventually the entirety of `act` is going to live in ReactFiberScheduler, behind a feature flag. That's probably what I would have done in this PR, in fact: Implement `act` behind a feature flag, and turn that on for test and noop renderers. For DOM you can use a (temporary) forked implementation. Then in a subsequent PR you can create a separate build of React DOM that turns on the feature flag.

threepointone replied:
Only for bytes, since dan and seb had mentioned they’d like to ship as little as possible in Reactdom. That’s why I’m using the hack to flush effects too; in a precious commit, most of this was in the FiberScheduler. I could move that back in and put it behind a feature flag ofc.

threepointone replied:
unsure how to proceed here. I could either move stuff into FiberScheduler, and add some weight to ReactDOM, or keep it outside, and copy paste across the 3 targets as I do now. and maybe I should put the feature flag in a follow up PR? I can bring this up on monday in chat.

acdlite commented on packages/react-dom/src/test-utils/ReactTestUtils.js:
```diff
@@ -390,58 +438,120 @@ const ReactTestUtils = {
+            result.then(
+              () => {
+                flushEffectsAndMicroTasks(doesHavePassiveEffects, () => {
+                  if (onDone !== undefined) {
```
Why is onDone ever undefined?

threepointone replied:
It’s not passed in prod (it comes from actedUpdates, which needs it only to decrement the depth and possibly issue a warning, neither of which we do in prod)

acdlite commented on packages/react-dom/src/test-utils/ReactTestUtils.js:
```diff
@@ -390,58 +438,120 @@ const ReactTestUtils = {
+                  if (onDone !== undefined) {
+                    onDone();
+                  }
+                  if (typeof successFn === 'function') {
```
I think we can assume `successFn` is a function. We do the same for Suspense thenables.

acdlite commented on packages/react-dom/src/test-utils/ReactTestUtils.js:
```diff
@@ -390,58 +438,120 @@ const ReactTestUtils = {
+                if (onDone !== undefined) {
+                  onDone(err);
+                }
+                errorFn(err);
```
Unlike with `successFn` above, you're not checking if `errorFn` exists. That's probably ok, because we expect people to use async/await, or wrap in Promise.resolve. But it should be consistent regardless.

threepointone replied:
Ah good catch. Will fix both. I think I had a test where I’d manually called .then. I’ll remove the checks.

acdlite commented:
> happy to export doesHavePassiveEffects as well btw, but didn’t want to stop my refactor in the meanwhile.

I would prefer it to be a separate export. It's always the exact same function, and passing it around everywhere makes the code pretty confusing to follow. More importantly, though, we are pretty strict about avoiding things like functions as values and dynamic values in this codebase, particularly for internal abstractions, which is what raises this above the level of a nit for me.

acdlite commented:
I added a test case in 04a4e91d0f1e1da581c4505813aa802ecdfbad60. It passes, and I assumed it would probably pass based on reading the implementation, but I wasn't positive until I wrote it. (It also wouldn't have passed in earlier iterations of this PR, which is why it's important to have for later refactors.) I only wrote it for test renderer, but we probably should have a corresponding one for each version of `act`, especially if the implementations are forked as they currently are.

threepointone commented:
I agree! I’m working on expanding the test coverage this weekend for the 3 targets.

acdlite commented:
Btw please only work on the weekend if/when it pleases you :D There's no pressure for you to do so otherwise (and if anyone does pressure you I'll ~beat them up~ give them a stern talking to).

threepointone commented:
Haha no, it’s an interesting problem, and I’m in too deep. I’m kinda having fun.

threepointone commented on packages/shared/createAct.js:
```diff
@@ -0,0 +1,162 @@
+
+export default function createAct(
+  actedUpdates: (() => void) => void,
+  // $FlowFixMe
```
I need to fix the type signature here

acdlite commented on packages/react-dom/src/test-utils/ReactTestUtilsAct.js:
```diff
@@ -0,0 +1,219 @@
+  flushPassiveEffects,
+] = ReactDOM.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.Events;
+
+const RendererHelpers = {
```
Why this extra wrapper?

threepointone replied:
Ok so I have a constraint that the 'rest' of the file should be copy paste-able across implementations. In the Noop renderer versions, because of a cyclical import, I cannot 'read' the values `flushPassiveEffects` and `batchedUpdates` off the renderer at module loading time (ie - `flushPassiveEffects` is not available on module start, so I have to do `ReactNoop.flushPassiveEffects()` instead of `flushPassiveEffects()` later). To unify the 2 constraints, I made this object, which shares its shape (for these 2 functions) with the renderer.  (So in the Noop renderer, I just do `import RendererHelpers from 'ReactNoop'`)Thought it was an elegant solution. Happy to hear if you have an alternate idea!

sebmarkbage replied:
The simplest solution to this is to just not have a separate file for ReactNoopAct. Just put it in the same file.

threepointone replied:
figured it was too clever by half ha. will do.

acdlite commented on packages/react/src/ReactSharedInternals.js:
```diff
@@ -15,6 +15,8 @@ import ReactDebugCurrentFrame from './ReactDebugCurrentFrame';
   ReactCurrentDispatcher,
   ReactCurrentOwner,
+  // used by act()
+  isActingUpdates: [false],
```
Let's use the same ref-style pattern as ReactCurrentOwner and ReactCurrentDispatcher:

```js
// Pick off the shared internals object
const {ReactShouldWarnActingUpdates} = ReactSharedInternals;

// Mutate like a ref
ReactShouldWarnActingUpdates.current = true;
```

acdlite commented on packages/react-reconciler/src/ReactFiberHooks.js:
```diff
@@ -1046,16 +1046,14 @@ function updateMemo<T>(
   // $FlowExpectedError - because requirements don't give a damn about your type sigs.
   if ('undefined' !== typeof jest) {
-    shouldWarnForUnbatchedSetState = true;
+    shouldWarnForUnactedSetState = true;
```
This doesn't need to be a module-level variable. Make this a local check right before firing the warning.

acdlite commented on packages/react-dom/src/test-utils/ReactTestUtilsAct.js:
```diff
@@ -0,0 +1,219 @@
+
+let didWarnAboutMessageChannel = false;
+
+let enqueueTask;
```
`enqueueTask` is the same across all these implementations, right? Might as well move it to `shared`.

acdlite replied:
The reason not to move it to `shared/` is if you're able to choose a more specific implementation per renderer. Like if we assume that the test renderer always runs in a Node environment.

threepointone replied:
I don't think we can make that guarantee. I'll move it into a `shared/` module.

acdlite commented on packages/react-dom/src/test-utils/ReactTestUtilsAct.js:
```diff
@@ -0,0 +1,219 @@
+  };
+}
+
+function createActedUpdatesScope(callback: (onDone: (?Error) => void) => void) {
```
What's the purpose of this closure? It looks like you're trying to support multiple, concurrent `act` sequences. I don't think we care to support that. And it wouldn't work as-is, anyway, because of the shared mutable state that lives on ReactSecretInternals.

acdlite replied:
Oh, is the closure merely so you can fire the "overlapping act() calls" warning below? You can use a module level variable to track that, instead of this wrapper.

threepointone replied:
the opposite, rather. because we don't want to support concurrent act sequences, we track the depth count changes, so we can warn if they do try to do so (eg- `act(async...); act(async...)` without `await`s, should warn)

acdlite commented on packages/react-noop-renderer/src/createReactNoop.js:
```diff
@@ -823,55 +818,15 @@ function createReactNoop(reconciler: Function, useMutation: boolean) {
-      NoopRenderer.updateContainer(null, root, null, null);
+    flushPassiveEffects: NoopRenderer.flushPassiveEffects,
+
+    // we patch in .act() after it's created
```
Why? My best guess is because you want to read `NoopRenderer.flushPassiveEffects` but that doesn't seem like a great reason.

threepointone replied:
right, or I'll have to put the implementation of .act() inline here inside createReactNoop, which I was looking to avoid.

acdlite replied:
https://github.com/facebook/react/pull/14853#discussion_r271445977

threepointone replied:
is this too clever? I don't mind moving the implementation in here tbh.

acdlite commented:
Remember that the endgame is to move all of this code into React itself, and use the host config to only include it in certain bundles like the noop and test renderers. In retrospect, I would have suggested that you do that in this PR since then you'd only need two implementations: the one that lives in React, and an extra one for React DOM for backwards compatibility (which we'll eventually remove). Since you're almost done and that approach requires futzing with the host config and build set up, don't worry about that for now; only bringing it up so to assuage your concerns that this will be difficult to maintain.

threepointone commented:
addressed most of the feedback, now moving the act() implementation back into createReactNoop

threepointone commented:
alright, done that too now.

acdlite commented on packages/react-reconciler/src/ReactFiberScheduler.old.js:
```diff
@@ -1836,9 +1842,20 @@ function scheduleWorkToRoot(fiber: Fiber, expirationTime): FiberRoot | null {
+    if (
+      ReactShouldWarnActingUpdates.current === false &&
+      isRendering === false &&
+      isBatchingUpdates === false
```
Let's check `isBatchingUpdates` first because that one will usually fail and short circuit the others. `ReactShouldWarnActingUpdates.current` check should come last because it's the most expensive.

threepointone commented:
71st time's the charm!

kentcdodds commented:
![yes gif](https://media.giphy.com/media/t3Mzdx0SA3Eis/giphy.gif)

Hooray! <3 Thanks everyone 👏

diegohaz commented:
🎉🎉🎉🎉

My hacky code is grateful for this amazing work.

![image](https://user-images.githubusercontent.com/3068563/55439146-12b91000-557a-11e9-9cea-fd08a3fd93f1.png)

gaearon commented on packages/shared/enqueueTask.js:
```diff
@@ -0,0 +1,42 @@
+  // assuming we're in node, let's try to get node's
+  // version of setImmediate, bypassing fake timers if any
+  let r = require; // trick packagers not to bundle this stuff.
+  enqueueTask = r('timers').setImmediate;
```
Do we know for sure webpack won't detect this / warn about this?
Especially if this turns into a `require` call with a GCC pass.

threepointone replied:
good point. I'll give it a spin later today/tomorrow

mrm007 replied:
Moment.js is using [the same method](https://github.com/moment/moment/blob/develop/src/lib/locale/locales.js#L55-L56) and **SPOILER ALERT** it doesn't work:

<img width="353" alt="image" src="https://user-images.githubusercontent.com/3297808/55454901-4acd5b00-562d-11e9-876a-bcf220fa640b.png">

threepointone replied:
I see. I have an idea to work around it, will try later today. But how do people use moment with webpack then?

Jessidhia replied:
The _actual_ thing you need is `__non_webpack_require__`

threepointone replied:
Aha, I see the problem, people futz around with plugins to exclude locales. Ok cool, i’ll try a workaround. Thanks for the heads up!

threepointone replied:
We might not be able to special case just for webpack. I remember trying this a while ago and it working `(() => eval(‘require’))`

threepointone replied:
To be clear, this is also a bit of an edge case, you’d have to be using webpack on your tests for this to be an issue. But will try to fix this before releasing.

threepointone replied:
(Great, the new guy gets to ship eval into react. Just great. 🍿)

threepointone replied:
Another option, if one’s running their tests through webpack, would be to mark react/react-dom as externals.

philipp-spiess replied:
fwiw that's how I outsmarted the compiler once:

```
export function requireAtRuntime(moduleName: string): any {
  const global = getGlobal()
  const require = global[atob('cmVxdWlyZQ==')]
  return require(moduleName)
}
```

gaearon replied:
We definitely shouldn't be adding special hacks for webpack alone.

gaearon replied:
Eval also won't work because it's restricted in some environments.

gaearon replied:
>if one’s running their tests through webpack

Does this have something to do with tests? I thought this file would be pulled in by all renderers. Regardless of whether you're writing tests or not.

gaearon replied:
Oh I see, react-dom doesn't pull it in.

gaearon replied:
I checked CRA, and seems like this isn't a problem there. Maybe it's ok to wait until somebody actually reports it.

gaearon replied:
In CRA this causes webpack's polyfill of `timers` to get bundled.

threepointone replied:
Could enable jest.useFakeTimers() and await an async act() call to see if it proceeds?

gaearon replied:
I don't understand what you mean

gaearon replied:
https://github.com/facebook/react/pull/15305

mpeyper commented:
@threepointone Is there any word on when [`16.9.0`](https://github.com/facebook/react/releases/tag/v16.9.0-alpha.0) will no longer be considered alpha?  It seems to be working fine.

ranneyd commented:
> @threepointone Is there any word on when [`16.9.0`](https://github.com/facebook/react/releases/tag/v16.9.0-alpha.0) will no longer be considered alpha? It seems to be working fine.

If you look at the logs this is, literally, the only new feature in 16.9.0. Not exactly enough for a new release

mpeyper commented:
> If you look at the logs this is, literally, the only new feature in 16.9.0.

Yep, that's why I linked to the release notes in my comment.

> Not exactly enough for a new release

I was not aware that there was a requisite number of features required for a release. Please, how many features, exactly, would be needed to get this non-breaking change released?

We're coming up on months since this was merged and many are eagerly awaiting it, but many are also not able to use alpha builds for production code.

davidje13 commented:
Note that due to https://github.com/facebook/react/issues/4542, using this alpha release causes a lot of NPM warnings when installing (even the alpha for react-dom is complaining about the alpha react!)

> ```
> npm WARN react-dom@16.9.0-alpha.0 requires a peer of react@^16.0.0 but none is installed. You must install peer dependencies yourself.
> ```

Because of these warnings, and because without this feature I'm having to skip tests (along with many other developers by the sounds of it), and because this feature appears to be stable, I strongly agree with @mpeyper that this should see an official release. I'd suggest if it doesn't feel like enough of a change for a 16.9 release, maybe 16.8.7 would make sense? This really feels like a bugfix for the incomplete `act` shipped in 16.8.

threepointone commented:
I understand the frustration, I'm sorry for it. We're working on some followup items that popped up, specifically for act() there's an umbrella item for them over here https://github.com/facebook/react/issues/15472 There are a couple of other bugfixes we want to get in, and we'll release as soon as it's ready.

kentcdodds commented:
![Driving GIF-downsized_large](https://user-images.githubusercontent.com/1500684/57570719-5d704800-73c2-11e9-9f9a-b8bee468e8de.gif)

😆
